src/codegen/generators/variable_generator.rs

```diff
@@ -98,6 +125,45 @@ impl<'ctx, 'b> VariableGenerator<'ctx, 'b> {
             );
         }
 
+        if let Some((location, format)) = &self.got_layout_file {
+
+            let got_entries = read_got_layout(location.as_str(), *format);
+            let mut new_globals = Vec::<String>::new();
+            let mut new_got_entries = HashMap::<String, u64>::new();
+            let mut new_got = HashMap::<u64, String>::new();
+
+            for (name, _) in &globals {
+                if let Some(idx) = got_entries.get(&name.to_string()) {
+                    new_got_entries.insert(name.to_string(), *idx);
+                    new_got.insert(*idx, name.to_string());
+                } else {
+                    new_globals.push(name.to_string());
+                }
+            }
+
+            // Put any globals that weren't there last time in any free space in the GOT.
+            let mut idx: u64 = 0;
+            for name in &new_globals {
+                while new_got.contains_key(&idx) {
+                    idx += 1;
+                }
+                new_got_entries.insert(name.to_string(), idx);
+                new_got.insert(idx, name.to_string());
+            }
+
+            // Now we can write new_got_entries back out to a file.
+            write_got_layout(new_got_entries, location.as_str(), *format);
+
+            // Construct our GOT as a new global array. We initialise this array in the loader code.
+            let got_size = new_got.keys().max().map_or(0, |m| *m + 1);
```

Comment:
Since the map contains indices, it's 1 + max index. So if the maximum index is 0, we have 1 in the array so we should create an array of size 1. I suppose we could avoid creating the array at all if it's empty, but also it might be useful in other parts of the system to be able to assume that `__custom_got` exists.